admittance_controller/include/admittance_controller/admittance_rule_impl.hpp

```diff
@@ -0,0 +1,534 @@
+// Copyright (c) 2022, PickNik, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+/// \authors: Denis Stogl, Andy Zelenak, Paul Gesel
+
+#ifndef ADMITTANCE_CONTROLLER__ADMITTANCE_RULE_IMPL_HPP_
+#define ADMITTANCE_CONTROLLER__ADMITTANCE_RULE_IMPL_HPP_
+
+#include <tf2_ros/transform_listener.h>
+#include "admittance_controller/admittance_rule.hpp"
+
+#include "rclcpp/duration.hpp"
+#include "rclcpp/utilities.hpp"
+
+namespace admittance_controller {
+
+  controller_interface::return_type
+  AdmittanceRule::configure(const std::shared_ptr<rclcpp_lifecycle::LifecycleNode> &node, int num_joints) {
+    // configure admittance rule using num_joints and load kinematics interface
+
+    num_joints_ = num_joints;
+
+    // initialize ROS functions
+    clock_ = node->get_clock();
+    tf_buffer_ = std::make_shared<tf2_ros::Buffer>(clock_);
+    tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);
+
+    // initialize memory and values to zero  (non-realtime function)
+    reset();
+
+    // Load the differential IK plugin
+    if (!parameters_->kinematics_.plugin_name_.empty()) {
+      try {
+        kinematics_loader_ = std::make_shared<pluginlib::ClassLoader<kinematics_interface::KinematicsBaseClass >>(
+            parameters_->kinematics_.plugin_package_, "kinematics_interface::KinematicsBaseClass");
+        kinematics_ = std::unique_ptr<kinematics_interface::KinematicsBaseClass>(
+            kinematics_loader_->createUnmanagedInstance(parameters_->kinematics_.plugin_name_));
+        if (!kinematics_->initialize(node, parameters_->kinematics_.tip_)) {
+          return controller_interface::return_type::ERROR;
+        }
+      }
+      catch (pluginlib::PluginlibException &ex) {
+        RCLCPP_ERROR(rclcpp::get_logger("AdmittanceRule"), "Exception while loading the IK plugin '%s': '%s'",
+                     parameters_->kinematics_.plugin_name_.c_str(), ex.what());
+        return controller_interface::return_type::ERROR;
+      }
+    } else {
+      RCLCPP_ERROR(rclcpp::get_logger("AdmittanceRule"),
+                   "A differential IK plugin name was not specified in the config file.");
+      return controller_interface::return_type::ERROR;
+    }
+
+    return controller_interface::return_type::OK;
+  }
+
+  controller_interface::return_type AdmittanceRule::reset() {
```

Comment:
I don't think I fully understand. I marked reset as non-real time because it reserves an unspecified amount of memory, since the number of joints is only known at runtime. The values are filled when the `parameter_handler_` is created [here ](https://github.com/pac48/ros2_controllers/blob/27ea60f5bdf6058c671f33a0c5607738eb914be7/admittance_controller/src/admittance_controller.cpp#L170)

Reply:
We should probably reset all the values before admittance controller get activated. And this each time the controller gets deactivated and activated again. `on_activate` should be real-time method, so I propose to reserve all the memory in the `on_configure`.

I understand that number of joints comes through parameters but is reserved memory and number of joints are not compatible we should force user to clean up and configure controller again. I am not convinced that we need to support changing number of joints of controller between two activation cycles.